### PR TITLE
[SPARK-27217][SQL] Nested column aliasing for more operators which can prune nested column

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

Currently we only push nested column pruning from a Project through a few operators such as LIMIT, SAMPLE, etc. There are a few operators like Aggregate, Expand which can prune nested columns by themselves, without a Project on top.

This patch extends the feature to those operators.

### Why are the changes needed?

Currently nested column pruning only applied on a few cases. It limits the benefit of nested column pruning. Extending nested column pruning coverage to make this feature more generally applied through different queries.

### Does this PR introduce _any_ user-facing change?

Yes. More SQL operators are covered by nested column pruning.

### How was this patch tested?

Added unit test, end-to-end tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 1]
```diff
@@ -68,10 +76,23 @@ object NestedColumnAliasing {
...
+      nestedFieldToAlias: Map[ExtractValue, Alias],
       attrToAliases: Map[ExprId, Seq[Alias]]): LogicalPlan = {
     plan.withNewChildren(plan.children.map { plan =>
       Project(plan.output.flatMap(a => attrToAliases.getOrElse(a.exprId, Seq(a))), plan)
-    })
+    }).transformExpressions {
+      case f: ExtractValue if nestedFieldToAlias.contains(f) =>
+        nestedFieldToAlias(f).toAttribute
+    }
+  }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:87]: Some change here is shared with #28556.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 2]
```diff
@@ -68,10 +76,23 @@ object NestedColumnAliasing {
...
+
+  /**
+   * Returns true for those operators that we can prune nested column on it.
+   */
+  private def canPruneOn(plan: LogicalPlan) = plan match {
+    case _: Aggregate => true
+    case _: Expand => true
+    case _ => false
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:95]: Are the current entries all the case we can support for nested column pruning? How about `FlatMapGroupsInPandas`?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:95]: I think FlatMapGroupsInPandas is also supported.

As you see, we do `transformExpressions` to replace `ExtractValue` in operator's expressions. So if the operator follows SparkSQL's operators, it should be fine.

This patch adds the change needed for general nested column pruning for the kind of operators which can prune nested column. Ideally we only need to add it in `canPruneOn`, and add test for it.

Currently I think nested column pruning test cases are all in Scala, no Python. I will think about how to add test.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:95]: okay, thanks for the check.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:95]: cc: @HyukjinKwon

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:95]: I think I was wrong. Re-checking `FlatMapGroupsInPandas`'s Python API, it looks like

```python
df.groupby("id").apply(normalize).show()
```

So basically the python udf takes no nested column selection but a full columns of DataFrame. It doesn't do nested column pruning.

**Review by viirya — COMMENTED** [review]
@maropu Thanks for reviewing. Do you have more comments on this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 3]
```diff
@@ -35,6 +35,11 @@ object NestedColumnAliasing {
     case Project(projectList, child)
         if SQLConf.get.nestedSchemaPruningEnabled && canProjectPushThrough(child) =>
       getAliasSubMap(projectList)
+
+    case plan if SQLConf.get.nestedSchemaPruningEnabled && canPruneOn(plan) =>
+      val exprsToPrune = plan.expressions
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:40]: nit: the consistent var name?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 4]
```diff
@@ -48,7 +53,10 @@ object NestedColumnAliasing {
     case Project(projectList, child) =>
       Project(
         getNewProjectList(projectList, nestedFieldToAlias),
-        replaceChildrenWithAliases(child, attrToAliases))
+        replaceChildrenWithAliases(child, nestedFieldToAlias, attrToAliases))
+
+    case other =>
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:58]: This case only matches `Aggregate` and `Expand` now?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala [hunk 5]
```diff
@@ -341,6 +341,78 @@ class NestedColumnAliasingSuite extends SchemaPruningTest {
       .analyze
     comparePlans(optimized, expected)
   }
+
+  test("Nested field pruning for Aggregate") {
+    val query1 = contact.groupBy($"id")(first($"name.first").as("first")).analyze
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:346]: Could you add tests for the cases, `contact.(limit/repartition/sample).groupBy()...`? IIUC this PR could support them, too?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala [hunk 6]
```diff
@@ -341,6 +341,78 @@ class NestedColumnAliasingSuite extends SchemaPruningTest {
...
+
+  test("Nested field pruning for Expand") {
+    val query1 = Expand(
+      Seq(
+        Seq($"name.first", $"name.middle"),
+        Seq(ConcatWs(Seq($"name.first", $"name.middle")),
+          ConcatWs(Seq($"name.middle", $"name.first")))
+      ),
+      Seq('a.string, 'b.string),
+      contact
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:379]: ditto; https://github.com/apache/spark/pull/28560#discussion_r428665197

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:58]: Yea, `canPruneOn` decides what operator will be pruned on.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala (hunk 5 above)
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:346]: ok, let me add some tests.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 4 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:58]: Ah, I see. Thanks for the check. If so, it might be better to explicitly filiter them here for readablity like `case other if canPruneOn(plan)`, or leave some comments about that. Both is okay to me.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:58]: Let me leave some comments, running `canPruneOn` twice seems redundant.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala [hunk 7]
```diff
@@ -341,6 +341,100 @@ class NestedColumnAliasingSuite extends SchemaPruningTest {
...
+        .select('name.getField("last").as(aliases2(0)), 'name.getField("first").as(aliases2(1)))
+      ).groupBy($"${aliases2(0)}")(first($"${aliases2(1)}").as("first")).analyze
+      comparePlans(optimized2, expected2)
+    }
+
+    Seq(
+      (plan: LogicalPlan) => plan,
+      (plan: LogicalPlan) => plan.limit(100),
+      (plan: LogicalPlan) => plan.repartition(100),
+      (plan: LogicalPlan) => Sample(0.0, 0.6, false, 11L, plan)).foreach {  base =>
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:372]: Thanks for this update and looks fine.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala [hunk 8]
```diff
@@ -338,6 +340,75 @@ abstract class SchemaPruningSuite
     }
   }
 
+  testSchemaPruning("select one deep nested complex field after repartition") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:343]: Is this test related to this PR?

**Comment by maropu** [comment]
Looks okay cc: @dongjoon-hyun @dbtsai

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala (hunk 8 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:343]: Not for this change, but we don't have test case for repartition. Added one to improve test coverage.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:343]: Ah, ok.

**Comment by viirya** [comment]
cc @cloud-fan

**Review by maropu — APPROVED** [review]

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
ping @cloud-fan @dongjoon-hyun

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:96]: Sorry I just checked this. Yes, I think ^ is correct.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 9]
```diff
@@ -204,15 +226,8 @@ object GeneratorNestedColumnAliasing {
       g: Generate,
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:226]: What about removing this method and call `NestedColumnAliasing.replaceChildrenWithAliases` directly above?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Let me merge this assuming the comment will be addressed at #28556.

**Comment by HyukjinKwon** [comment]
Merged to master.
